File: CPP/Algorithm/Factorial.cpp
// Complexity: O(n)
// Iterative Approach:  // Results till 20!

int64_t fac(int64_t n) {
    int64_t res = 1;
    for (int i = 2; i <= n; ++i) res *= i;
    return res;
}

// Taking the values modulo M = 1e9 + 7:

int64_t fac(int64_t n) {
    int64_t res = 1;
    for (int i = 2; i <= n; ++i) res = (res * i) % M;
    return res;
}

// Recursive Approach:

int64_t fac(int64_t n) {
    return (n ? (fac(n - 1) * n) : 1);
}

// Storing the factorial values:

int64_t fac[N];
void pre() {
    fac[0] = fac[1] = 1;
    for (int i = 2; i < N; ++i) fac[i] = (fac[i - 1] * i) % M;
}

// String Approach:  // O(n^2)

auto fac(int64_t n) {
    string s = "1";
    for (int64_t i = 2; i <= n; ++i) {
        int64_t c = 0, sz = s.size();
        for (int j = 0; j < sz; ++j) {
            int64_t a = (s[j] - 48) * i + c; s[j] = (a % 10) + 48; c = a / 10;
        }
        while (c) {
            s += ((c % 10) + 48); c /= 10;
        }
    }
    reverse(s.begin(), s.end()); return s;
}

// Vector Approach:

auto fac(int64_t n) {
    vector<int64_t> v = {1};
    for (int64_t i = 2; i <= n; ++i) {
        int64_t c = 0, sz = v.size();
        for (int j = 0; j < sz; ++j) {
            int64_t a = v[j] * i + c; v[j] = a % 10; c = a / 10;
        }
        while (c) {
            v.emplace_back(c % 10); c /= 10;
        }
    }
    reverse(v.begin(), v.end()); return v;
}

File: Python/Algorithm/Factorial.py
# Complexity: O(n)

def fac(n):
    res = 1
    for i in range(2, n + 1): res *= i
    return res

def fac(n):
    res = 1
    for i in range(2, n + 1): res = (res * i) % M
    return res

# Recursive Approach:

def fac(n): return 1 if n == 0 else n * fac(n - 1)

# Storing the fac values:

fac = [1] * N
for i in range(2, N): fac[i] = fac[i - 1] * i % M

# String Approach: O(n^2)

def fac(n):
    s = "1"
    for i in range(2, n + 1):
        car = 0
        s = [int(digit) for digit in s]
        for j in range(len(s)):
            a = s[j] * i + car
            s[j] = a % 10; car = a // 10
        while car: s.append(car % 10); car //= 10
        s = ''.join(map(str, s[::-1]))
    return s

# Vector Approach:

def fac(n):
    v = [1]
    for i in range(2, n + 1):
        car = 0
        for j in range(len(v)):
            a = v[j] * i + car
            v[j] = a % 10; car = a // 10
        while car: v.append(car % 10); car //= 10
        v.reverse()
    return v